File: app/src/main/java/com/example/alarm/room/AlarmDao.java
package com.example.alarm.room;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;


@Dao
public interface AlarmDao {
    @Query("SELECT * FROM alarms ORDER BY time_hours ASC, time_minutes ASC")
    List<AlarmDbEntity> getAlarms();

    @Query("SELECT COUNT(*) FROM alarms WHERE time_hours = :hours AND time_minutes = :minutes")
    int countAlarmsWithTime(int hours, int minutes);

    @Query("SELECT COUNT(*) FROM alarms WHERE time_hours = :hours AND time_minutes = :minutes AND name = :name")
    int countAlarmsWithTimeAndName(int hours, int minutes, String name);

    @Insert
    void addAlarm(AlarmDbEntity alarmDbEntity);

    @Query("UPDATE alarms SET enabled = :enabled WHERE id = :id")
    void updateEnabled(Long id, Boolean enabled);

    @Update
    void updateAlarm(AlarmDbEntity alarmDbEntity);

    @Delete(entity = AlarmDbEntity.class)
    void deleteAlarm(AlarmDbEntity alarmDbEntity);
}

File: app/src/main/java/com/example/alarm/room/AlarmDao.kt
package com.example.alarm.room

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Update
import kotlinx.coroutines.flow.Flow

@Dao
interface AlarmDao {
    @Query("SELECT * FROM alarms ORDER BY time_hours ASC, time_minutes ASC")
    suspend fun selectAlarms(): List<AlarmsGetTuple?>

    @Query("SELECT COUNT(*) FROM alarms WHERE time_hours = :hours AND time_minutes = :minutes")
    suspend fun countAlarmsWithTime(hours: Int, minutes: Int): Int

    @Insert
    suspend fun addAlarm(alarmDbEntity: AlarmDbEntity)

    @Update(entity = AlarmDbEntity::class)
    suspend fun updateEnabled(updateEnabledTuple: AlarmUpdateEnabledTuple)

    @Update
    suspend fun updateAlarm(alarmDbEntity: AlarmDbEntity)

    @Query("SELECT * FROM alarms WHERE id = :alarmId")
    fun getById(alarmId: Long): Flow<AlarmDbEntity?>

    @Delete(entity = AlarmDbEntity::class)
    suspend fun deleteAlarm(alarmDbEntity: AlarmDbEntity)
}